From 7a94dae70fad11d06f7e388f886386f0b0c0f436 Mon Sep 17 00:00:00 2001
From: Antonio Borneo <antonio.borneo@foss.st.com>
Date: Tue, 21 Jan 2025 12:29:12 +0100
Subject: [PATCH 1575/1695] drivers: stm32_exti: add helper
 stm32_exti_get_pdata()

Add temporarily a new helper for the transition to irq framework.
It will be removed once the consumer devices with get the handler
from the generic interrupt framework.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git f522bb6b4732e4e7f6726a4818f1b6215b983431 drivers: stm32_exti: add helper stm32_exti_get_pdata()]
Signed-off-by: Antonio Borneo <antonio.borneo@foss.st.com>
Change-Id: Ifddc44176a09ae3a617082e88d593f28523e59a5
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/434922
Reviewed-by: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Domain-Review: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 .../plat-stm32mp1/drivers/stm32mp1_pwr_irq.c  |  5 +----
 .../plat-stm32mp2/drivers/stm32mp25_pwr_irq.c |  5 +----
 core/drivers/stm32_exti.c                     |  8 ++++++++
 core/drivers/stm32_rtc.c                      |  5 +----
 core/drivers/stm32_tamp.c                     |  5 +----
 core/include/drivers/stm32_exti.h             | 20 +++++++++++++++++++
 6 files changed, 32 insertions(+), 16 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr_irq.c b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr_irq.c
index a84e4f97c..259f79e0b 100644
--- a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr_irq.c
+++ b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr_irq.c
@@ -400,10 +400,7 @@ stm32mp1_pwr_irq_probe(const void *fdt, int node,
 	priv->base = stm32_pwr_base();
 
 	if (IS_ENABLED(CFG_STM32_EXTI)) {
-		res = dt_driver_device_from_node_idx_prop("wakeup-parent",
-							  fdt, node, 0,
-							  DT_DRIVER_INTERRUPT,
-							  &priv->exti);
+		res = stm32_exti_get_pdata(fdt, node, &priv->exti);
 		if (res)
 			goto err;
 	}
diff --git a/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr_irq.c b/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr_irq.c
index 744c35ad0..904440b8a 100644
--- a/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr_irq.c
+++ b/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr_irq.c
@@ -392,10 +392,7 @@ stm32mp25_pwr_irq_probe(const void *fdt, int node, int interrupt)
 	priv = pwr_wkup_d;
 
 	if (IS_ENABLED(CFG_STM32_EXTI)) {
-		res = dt_driver_device_from_node_idx_prop("wakeup-parent",
-							  fdt, node, 0,
-							  DT_DRIVER_INTERRUPT,
-							  &priv->exti);
+		res = stm32_exti_get_pdata(fdt, node, &priv->exti);
 		if (res)
 			goto err;
 	}
diff --git a/core/drivers/stm32_exti.c b/core/drivers/stm32_exti.c
index ef73817ba..f682b63a3 100644
--- a/core/drivers/stm32_exti.c
+++ b/core/drivers/stm32_exti.c
@@ -527,6 +527,14 @@ stm32_exti_pm(enum pm_op op, unsigned int pm_hint,
 	return TEE_SUCCESS;
 }
 
+TEE_Result stm32_exti_get_pdata(const void *fdt, int nodeoffset,
+				struct stm32_exti_pdata **exti)
+{
+	return dt_driver_device_from_node_idx_prop("wakeup-parent",
+						   fdt, nodeoffset, 0,
+						   DT_DRIVER_INTERRUPT, exti);
+}
+
 static TEE_Result
 stm32_exti_get_handle(struct dt_pargs *pargs __unused, void *data,
 		      struct stm32_exti_pdata **out_data)
diff --git a/core/drivers/stm32_rtc.c b/core/drivers/stm32_rtc.c
index f8ea664c5..e45bf1147 100644
--- a/core/drivers/stm32_rtc.c
+++ b/core/drivers/stm32_rtc.c
@@ -795,10 +795,7 @@ static TEE_Result parse_dt(const void *fdt, int node)
 	}
 
 	if (rtc_dev.rtc->is_wakeup_source) {
-		res = dt_driver_device_from_node_idx_prop("wakeup-parent",
-							  fdt, node, 0,
-							  DT_DRIVER_INTERRUPT,
-							  &rtc_dev.exti);
+		res = stm32_exti_get_pdata(fdt, node, &rtc_dev.exti);
 		if (res == TEE_ERROR_ITEM_NOT_FOUND) {
 			EMSG("DT property 'wakeup-source' requires 'wakeup-parent'");
 			return res;
diff --git a/core/drivers/stm32_tamp.c b/core/drivers/stm32_tamp.c
index 927a8a871..c15601ea8 100644
--- a/core/drivers/stm32_tamp.c
+++ b/core/drivers/stm32_tamp.c
@@ -2016,10 +2016,7 @@ static TEE_Result stm32_tamp_parse_fdt(struct stm32_tamp_platdata *pdata,
 	parse_bkpregs_dt_conf(pdata, fdt, node);
 
 	if (pdata->is_wakeup_source && IS_ENABLED(CFG_STM32_EXTI)) {
-		res = dt_driver_device_from_node_idx_prop("wakeup-parent",
-							  fdt, node, 0,
-							  DT_DRIVER_INTERRUPT,
-							  &pdata->exti);
+		res = stm32_exti_get_pdata(fdt, node, &pdata->exti);
 		if (res == TEE_ERROR_DEFER_DRIVER_INIT)
 			return TEE_ERROR_DEFER_DRIVER_INIT;
 		if (res)
diff --git a/core/include/drivers/stm32_exti.h b/core/include/drivers/stm32_exti.h
index 3144ccab7..d7e6a6c94 100644
--- a/core/include/drivers/stm32_exti.h
+++ b/core/include/drivers/stm32_exti.h
@@ -10,6 +10,26 @@
 
 struct stm32_exti_pdata;
 
+/*
+ * Recover the pointer to stm32_exti_pdata from DT property
+ * "wakeup-parent"
+ * @fdt: FDT base address
+ * @nodeoffset: node offset in the FDT
+ * @exti: the EXTI private data
+ */
+#ifdef CFG_STM32_EXTI
+TEE_Result stm32_exti_get_pdata(const void *fdt, int nodeoffset,
+				struct stm32_exti_pdata **exti);
+#else
+static inline
+TEE_Result stm32_exti_get_pdata(const void *fdt __unused,
+				int nodeoffset __unused,
+				struct stm32_exti_pdata **exti __unused)
+{
+	return TEE_ERROR_NOT_IMPLEMENTED;
+}
+#endif
+
 /*
  * Set EXTI type in RTSR and FTSR EXTI registers
  * @exti_line: EXTI line number
-- 
2.47.2

